File: sdks/dotnet/src/SkMcp.AspNetCore/Discovery/RouteGlob.cs
using System.Text.RegularExpressions;

namespace SkMcp.AspNetCore.Discovery;

internal static class RouteGlob
{
    /// <summary>A <c>*</c> matches within one segment, <c>**</c> across segments.</summary>
    /// <remarks>
    /// Every regex metacharacter outside the two wildcards is escaped before it can turn a
    /// host-written pattern into a catastrophic backtracker; the timeout bounds what escaping
    /// cannot. Route templates carry <c>{id}</c> placeholders, so the braces reach that escape on
    /// every ordinary pattern.
    /// </remarks>
    public static bool Matches(string pattern, string route)
    {
        string source = string.Join(".*", pattern.Split("**")
            .Select(part => string.Join("[^/]*", part.Split('*').Select(Regex.Escape))));
        return Regex.IsMatch(
            route, $"^{source}$", RegexOptions.None, TimeSpan.FromSeconds(1));
    }
}

File: packages/http/core/src/route-glob.ts
/**
 * Matches a route template against a host-written pattern: `*` within one segment, `**` across.
 *
 * @param pattern - The host's pattern, matched whole; `**` is the catch-all, `*` is not.
 * @param route - A composed route template, as `EndpointDescriptor.route` carries it.
 * @returns Whether the pattern matches the whole route.
 */
export function matchesRoute(pattern: string, route: string): boolean {
  /**
   * Every regex metacharacter outside the two wildcards is escaped before it can turn a
   * host-written pattern into a catastrophic backtracker. Route templates carry `{id}`
   * placeholders, so the braces reach this escape on every ordinary pattern.
   */
  const source = pattern
    .split("**")
    .map((part) =>
      part
        .split("*")
        .map((literal) => literal.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"))
        .join("[^/]*"),
    )
    .join(".*");
  return new RegExp(`^${source}$`).test(route);
}
